File: sync/src/main/java/com/inceptionnotes/sync/World.java
package com.inceptionnotes.sync;

import com.inceptionnotes.sync.events.SyncEvent;
import com.inceptionnotes.sync.objects.Note;
import com.inceptionnotes.sync.store.NoteStore;

import java.util.ArrayList;
import java.util.Set;
import java.util.concurrent.ConcurrentHashMap;

/**
 * Created by jacob on 2/10/18.
 */

public class World {

    private final NoteStore noteStore = new NoteStore();
    private final Set<Client> clients = ConcurrentHashMap.newKeySet();

    public void join(Client client) {
        clients.add(client);
    }

    public void leave(Client client) {
        clients.remove(client);
    }

    public void noteChanged(Note note, Client culprit) {
        SyncEvent syncEvent = new SyncEvent();
        syncEvent.notes = new ArrayList<>();
        syncEvent.notes.add(note);

        clients.forEach(c -> {
            if (!c.isIdentified() || c.getShow() == null) {
                return;
            }

            if (!note.getId().equals(c.getShow()) && !noteStore.noteVisibleFromEye(c.getShow(), note.getId())) {
                return;
            }

            if (culprit == c) {
                return;
            }

            c.send(syncEvent);
            note.toSyncNote().getSync().forEach(prop ->
                noteStore.setPropSeenByClient(c.getClientId(), note.getId(), prop));
        });
    }
}

File: sync/src/main/java/com/inceptionnotes/sync/world/World.kt
package com.inceptionnotes.sync.world

import com.inceptionnotes.sync.events.SyncEvent
import com.inceptionnotes.sync.objects.Note
import com.inceptionnotes.sync.store.NoteStore
import com.queatz.on.On
import java.util.concurrent.ConcurrentHashMap

/**
 * Created by jacob on 2/10/18.
 */

class World(private val on: On) {

    private val clients = ConcurrentHashMap.newKeySet<Client>()

    fun join(client: Client) = synchronized(clients) { clients.add(client) }

    fun leave(client: Client) = synchronized(clients) { clients.remove(client) }

    fun onNotesChanged(notes: List<Note>, culprit: Client) = synchronized(clients) {
        clients.forEach { client ->
            if (!client.isIdentified || client.show == null) {
                return@forEach
            }

            if (culprit === client) {
                return@forEach
            }

            val syncEvent = SyncEvent()

            notes.forEach { note ->
                if (note.id == client.show || on<NoteStore>().noteVisibleFromEye(client.show!!, note.id!!)) {
                    syncEvent.notes.add(note)
                }
            }

            client.send(syncEvent)

            syncEvent.notes.forEach { note ->
                note.toSyncNote().sync?.forEach {
                    prop -> on<NoteStore>().setPropSeenByClient(client.clientId!!, note.id!!, prop)
                }
            }
        }
    }
}
